Remove Tensor(const Tensor&, BaseContext*, type)

dzhulgakov opened the pull request:
Differential Revision: D11915764

bddppq commented:
rocm tests:
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger-test/3739/
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger/3728/

Commit:
Remove calls to CopyFrom that can be sync

Summary:
CopyFrom without context argument does the sync copy on the current gpu - exactly what most of the places need.

This diff kills about 60% of CopyFrom usages. Most common pattern is gpu->cpu copy with further FinishDeviceComputation - the latter can be just killed.

Differential Revision: D11236076

fbshipit-source-id: c3539a02a69cd0cf1f5689d197075e46fdf16fd4

Commit:
Remove Tensor(const Tensor&, BaseContext*, type)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/13204

Differential Revision: D11915764

fbshipit-source-id: 03c4bae3838fbb679d42f809874e7586065ff972

ezyang commented on aten/src/ATen/core/TensorImpl.h:
```diff
@@ -847,6 +847,8 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
    * a tensor on CPU and then CopyFrom a CUDA tensor, that will to a
    * CUDA-to-CPU transfer).
+   *
+   * If the function is invoked without `context` the copy would be synchronous
```
I'm actually not sure how this follows from the definition of this method below. Don't you need `FinishDeviceComputation()` to synchronize the copy? I don't see this method call in the body below.

dzhulgakov replied:
Ah! The destructor of Context object calls FinishDeviceComputation. Context is basically RAII if it's created on the fly :)

ezyang replied:
Great, thank you :)

ezyang reviewed (approved):
Approving because I assume you know what you're doing, but I would be quite interested to know why the sync call is no longer necessary.

dzhulgakov commented:
@ezyang - it basically replaces two  patterns of usage:
1) CPUContext only, where passing or not passing the context is the same (and it's cleaner to not pass)
2) CPU->GPU copy where people actually wanted to do a sync copy just after and invoke context.FinishDeviceComputation explicitly.

And yes, current logic is convoluted - that's why I'm untangling it
